test/hotspot/jtreg/serviceability/jvmti/events/ClassLoad/classload01/libclassload01.cpp

```diff
@@ -0,0 +1,240 @@
+#include <stdlib.h>
+#include <string.h>
+
+#include <jvmti.h>
+
+#include "jvmti_common.h"
+
+extern "C" {
+
+#define PASSED  0
+#define STATUS_FAILED  2
+
+/* classes which must have the class load event */
+static const char *expSigs[] = {
+    "Lclassload01;",
+    "Lclassload01$TestedClass;",
+    "LTestedClassVirtual;",
+};
+#define EXP_SIG_NUM (sizeof(expSigs)/sizeof(char*))
+
+static const jboolean expThreadIsVirtual[] = {
+    JNI_FALSE,
+    JNI_FALSE,
+    JNI_TRUE
+};
+
+/* classes which must not have the class load event */
+static const char *unexpSigs[] = {
+    "Z", /* boolean */
+    "B", /* byte */
+    "C", /* char */
+    "D", /* double */
+    "F", /* float */
+    "I", /* integer */
+    "J", /* long */
+    "S", /* short */
+
+    "[Z", /* boolean array */
+    "[B", /* byte array */
+    "[C", /* char array */
+    "[D", /* double array */
+    "[F", /* float array */
+    "[I", /* integer array */
+    "[J", /* long array */
+    "[S", /* short array */
+    "[Lclassload01$TestedClass;"
+};
+#define UNEXP_SIG_NUM (sizeof(unexpSigs)/sizeof(char*))
+
+static volatile int clsEvents[EXP_SIG_NUM];
+static volatile int primClsEvents[UNEXP_SIG_NUM];
+
+static jint result = PASSED;
+static jvmtiEnv *jvmti = NULL;
+static jvmtiEventCallbacks callbacks;
+static jrawMonitorID counter_lock;
+
+static void initCounters() {
+  size_t i;
+
+  for (i = 0; i < EXP_SIG_NUM; i++) {
+    clsEvents[i] = 0;
+  }
+
+  for (i = 0; i < UNEXP_SIG_NUM; i++) {
+    primClsEvents[i] = 0;
+  }
+}
+
+static int findSig(char *sig, int expected) {
+  unsigned int i;
+
+  for (i = 0; i < ((expected == 1) ? EXP_SIG_NUM : UNEXP_SIG_NUM); i++) {
+    if (sig != NULL &&
+        strcmp(((expected == 1) ? expSigs[i] : unexpSigs[i]), sig) == 0) {
+      return i; /* the signature found, return index */
+    }
+  }
+  return -1; /* the signature not found */
+}
+
+/** callback functions **/
+void JNICALL
+ClassLoad(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread, jclass klass) {
+  int i = 0;
+  char *sig, *generic;
+  jvmtiError err;
+
+  RawMonitorLocker rml(jvmti, jni, counter_lock);
+
+  err = jvmti->GetClassSignature(klass, &sig, &generic);
+  if (err != JVMTI_ERROR_NONE) {
+    result = STATUS_FAILED;
+    LOG("TEST FAILURE: unable to obtain a class signature. Error %d\n", err);
+    return;
+  }
+
+  i = findSig(sig, 1);
+  if (i != -1) {
+    jboolean is_virtual_thread = jni->IsVirtualThread(thread);
+    print_thread_info(jvmti, jni, thread);
+    if (is_virtual_thread != expThreadIsVirtual[i]) {
+      LOG("TEST FAILED: IsVirtualThread(thread) is not expected: %d\n", is_virtual_thread);
+      result = STATUS_FAILED;
+    } else {
+      clsEvents[i]++;
+      LOG("CHECK PASSED: ClassLoad event received for the class \"%s\" as expected\n", sig);
+    }
+  } else {
+    i = findSig(sig, 0);
+    if (i != -1) {
+      result = STATUS_FAILED;
+      primClsEvents[i]++;
+      LOG(
+          "TEST FAILED: JVMTI_EVENT_CLASS_LOAD event received for\n"
```

Comment:
Combine 138+138.
I won't comment this more in hope the same will be fixed in all tests.